pkg/bgpv1/gobgp/server.go

```diff
@@ -91,38 +92,71 @@ func NewGoBGPServerWithConfig(ctx context.Context, log *logrus.Entry, params typ
 // AddNeighbor will add the CiliumBGPNeighbor to the gobgp.BgpServer, creating
 // a BGP peering connection.
 func (g *GoBGPServer) AddNeighbor(ctx context.Context, n types.NeighborRequest) error {
+	peer, err := g.getPeerForNeighbor(n)
+	if err != nil {
+		return err
+	}
+	peerReq := &gobgp.AddPeerRequest{
+		Peer: peer,
+	}
+	if err = g.server.AddPeer(ctx, peerReq); err != nil {
+		return fmt.Errorf("failed while adding peer %v %v: %w", n.Neighbor.PeerAddress, n.Neighbor.PeerASN, err)
+	}
+	return nil
+}
+
+// UpdateNeighbor will update the existing CiliumBGPNeighbor in the gobgp.BgpServer.
+func (g *GoBGPServer) UpdateNeighbor(ctx context.Context, n types.NeighborRequest) error {
+	peer, err := g.getPeerForNeighbor(n)
+	if err != nil {
+		return err
+	}
+	peerReq := &gobgp.UpdatePeerRequest{
+		DoSoftResetIn: true, // should perform soft reset only if needed
+		Peer:          peer,
+	}
+	if _, err = g.server.UpdatePeer(ctx, peerReq); err != nil {
+		return fmt.Errorf("failed while updating peer %v %v: %w", n.Neighbor.PeerAddress, n.Neighbor.PeerASN, err)
+	}
+	return nil
+}
+
+func (g *GoBGPServer) getPeerForNeighbor(n types.NeighborRequest) (*gobgp.Peer, error) {
```

Comment:
I tried this PR locally and noticed that changing any timer parameter hard resets the BGP peer. I think we're hitting this condition.

https://github.com/osrg/gobgp/blob/2fbfee5f2748e387e4442164760a6c56c1cc0578/pkg/server/server.go#L3289

I guess we need to copy the running configuration before the update? Could you check that?

Reply:
Will check that..

Reply:
Changing timer parameter didn't reset the connection, which makes sense to avoid user's surprise. Probably the same for GoBGP 🤔 thi

Reply:
I'd like to understand what should be expectation here. Changing neighbour parameters in CRD means user want this new configuration - and should expect peering to flap in that case - since some config is negotiated, right ? 

Or do we need to provide some sort of API where user can clear BGP sessions ( to re negotiate peering ) after CRD has been modified ?

Reply:
So what we figured out today is that even FRR does not reset peering if holdTime or keepAlive timers params change - even if holdTime is negotiated during session setup. Instead it would keep the new values as "configured" values and use them after next peer reset, e.g.:
```
  Hold time is 12, keepalive interval is 4 seconds
  Configured hold time is 9, keepalive interval is 3 seconds
```
GoBGP behaves the same, just the update needs to be done with a bit more care - we will dump the peer before updating, as many peer config fields are defaulted internally in GoBGP and would cause reset if not provided on update.

Then we will have a similar distinction in our state API, e.g. `applied-hold-time-seconds` and `configured-hold-time-seconds`. "Configured" values will be used only when the session resets - at the moment the only reset API for us is agent restart (or peer restart), but we may add an explicit API for that later.

Reply:
Implemented as described in my previous comment.